lib/api/streaming.js

```diff
@@ -178,6 +211,25 @@ Streaming.prototype.subscribe = function(name, listener) {
   return fayeClient.subscribe(channelName, listener);
 };
 
+/**
+ * Subscribe topic/channel
+ *
+ * @param {String} name - Topic name
+ * @param {Array} Extensions - Extensions to apply to subscription
+ * @param {Callback.<Streaming~StreamingMessage>} listener - Streaming message listener
+ * @returns {Subscription} - Faye subscription object
+ */
+Streaming.prototype.subscribeWithExtension = function(name, extensions, listener) {
+  var channelName = name.indexOf('/') === 0 ? name : '/topic/' + name;
+  var fayeClient = this._getFayeClient(channelName);
+
+  extensions.forEach(function(extension) {
+    fayeClient.addExtension(extension);
```

Comment:
It seems registering extensions to faye client instance, though the instance is shared with different subscribers which might not have any extensions.

Reply:
Thank you for your review. Yes, I see what you mean 🤔

As you know, at the heart of `Streaming` is an object that caches singleton Faye clients for the app. How can we cache these new Faye clients with extensions? Given the way extensions are complex objects containing anonymous functions, they cannot be identified distinctly in this cache. (Javascript woes 😩.) Without these distinctly cached Faye clients, each subscription will create a new Faye client, and the jsforce `unsubscribe()` functions will not work.

**Potential solution to avoid these side-effects**: we could create & return new Faye clients from `Streaming.prototype.subscribeWithExtension`. The main problems here are that an app that creates multiple streaming subscriptions with extensions will actually be running a Faye client for each subscription. Fortunately, the Faye Subscription objects returned from `subscribe()` calls do contain a `cancel()` function for unsubscribe.

I'm not sure how to best solve this. The jsforce `Streaming` design abstracts (hides) the underlying Faye clients. Now, adding extensions to them means that Faye clients cannot be abstracted the same way anymore. The extended Faye clients must be manually managed by the developer.

**Potential new design to avoid these side-effects**: replace this pull request's `subscribeWithExtensions()` with a lower-level interface that requires the developer to create the Faye client with extensions, and then use the Faye client directly, for example:

```javascript
const fayeClient = conn.streaming.withExtensions([authFailureExt]);
fayeClient.subscribe(channelName, listener);
```

I am in favor of this solution, as the lower-level Faye client interface is more natural anyway, once you're working with Faye extensions!

What do you think @stomita ? 😄

Reply:
Here's a full example, for how I envision this `withExtensions` design would enable Replay & Auth Failure handling in a server-side Node.js app:

```javascript
// Establish a Salesforce connection. (Details elided)
const conn = new jsforce.Connection({ … });

const channel = "/event/My_Event__e";
const replayId = -2; // -2 is all retained events

const exitCallback = () => process.exit(1);
const authFailureExt = new jsforce.Streaming.authFailureExtension(exitCallback);

const replayExt = new jsforce.Streaming.replayExtension(channel, replayId);

// Create a new Faye client
const fayeClient = conn.streaming.withExtensions([
  authFailureExt,
  replayExt
]);

// Then use Faye's API directly
const subscription = fayeClient.subscribe(channel, data => {
  console.log('topic received data', data);
});

// Sometime later, unsubscribe.
subscription.cancel();
```

If this seems like a satisfactory solution, I will commit this big revision soon!